### ISSUE TITLE
SVG elements don’t return offset values

### ISSUE THREAD

**Issue by arthurkirkosa** [jlmakes/scrollreveal#264]
I have an inline svg and I've applied classes to some groups.

After initializing ScrollReveal on the classes, the DOM is altered, but when scrolling over the svg (which is `100vh`, `100vw`, the animation is not executed; the script is not returning true to `_isElemVisible(elem)`

``` html
<g class="reveal" data-sr-id="1" style="; visibility: visible;  -webkit-transform: translateX(-5rem) scale(0.9); opacity: 0;transform: translateX(-5rem) scale(0.9); opacity: 0;">
  <path class="st1" d="M406.4 548.6h673.4c3.1 0 6.2.2 9.3.5v-.5l167.5 104.6-167.5 104.6v-.5c-3.1.3-6.2.5-9.3.5H406.4c-53.4 0-96.7-46.8-96.7-104.6s43.3-104.6 96.7-104.6z"></path>
</g>
```

``` js
var sr = ScrollReveal({
    distance: '5rem',
    duration: 1000,
    delay: 500,
    viewOffset: {
        top: 100
    }
});

sr.reveal('.reveal', { origin: 'left' });
```

OSX, Chrome Version 51.0.2704.84 (64-bit)

**Comment by jlmakes** [comment]
Hmm.

SVG’s have been problematic in the past (#73 #166). Any chance you could link me to a live example? I’ll be much better equipped to help you troubleshoot.

**Comment by arthurkirkosa** [comment]
[SVG on Dropbox](https://www.dropbox.com/s/vecz8bd4qpmw5wv/flower.svg?dl=0)

I'd like to animate the outer arrows; (.Reveal--left, .Reveal--right)

**Comment by jlmakes** [comment]
Thanks, but that is just an image. If I am to help you debug, I need _sample code_. I recommend [JSBin](http://jsbin.com)) but even a live website will do.

**Comment by arthurkirkosa** [comment]
http://jsbin.com/sozozuzoqa/edit?js,output

**Comment by jlmakes** [comment]
**tl;dr** — _ScrollReveal does not support SVG elements._

Next time please format the code a little nicer for your fellow developer, I had to add whitespace back into your SVG; the formatted code example: http://jsbin.com/jayima/edit?html,js,output

#### Debugging

I replaced the CDN link to ScrollReveal, with a debugger version that has all methods public. As it turns out, I believe the problem is this:

When ScrollReveal attempts to get the offset top, left, width and height of the `<g>` element — it’s returning `{ top: 0, left: 0, height: undefined, width: undefined }`. This means ScrollReveal has no notion of where or how big the element is, meaning it never qualifies as "visible" and is never revealed.

I believe the reason is this:

>##### SVGElement.offset(Parent|Top|Left|Width|Height)
"These attributes should only be supported on HTMLElement, but have been supported on all Elements in WebKit and Blink. getBoundingClientRect() can typically be used as a replacement."
**§** https://www.chromestatus.com/features/5724912467574784

#### Broken Everywhere

It doesn’t work in Safari, or Firefox either — returning the same 0 and undefined values. _If it worked in Chrome before version 50, it wasn’t supposed to!_

#### Solution?

The problem is that `getBoundingClientRect()` is the modern alternative to offsets, and does indeed work with SVG elements — but it includes CSS transforms in the returned coordinates, meaning when ScrollReveal moves something 150px to the right (before a reveal for example), the reported coordinates of that element include that 150px transformation... making it more challenging to determine the elements true visibility.

I actually tested using `getBoundingClientRect()` as the source of truth for version 3, but discovering this behavior was the reason I gutted it. (The result was reveal animations would jitter, revealing in/out over and over… never sure if they were visible or not.)

**Comment by arthurkirkosa** [comment]
> Next time please format the code a little nicer for your fellow developer, I had to add whitespace back into your SVG;

Sorry... had the svg minified.

**Comment by jlmakes** [comment]
No worries @arthurkirkosa, just wanted to mention it for next time — make it as easy as possible for someone to help :)

**Comment by jlmakes** [comment]
>"Is there a chance to animate SVG elements using ScrollReveal?"
— @anoopd

As you can see above, there are issues working with SVG elements directly. A solution you can use today is to wrap those SVG’s in a div, and use ScrollReveal to reveal that div.

(That is the technique used for the [ScrollReveal demo](https://scrollrevealjs.org).)

**Comment by jlmakes** [comment]
I revisited this issue for ScrollReveal 4, but again found no success when trying to use `getBoundingClientRect()` as the foundation for "element position checking" (EPC).

The fact that modern EPC such as `getBoundingClientRect()` and `IntersectionObserver` both include CSS Transforms in their return values, means the results vary as elements move/animate using CSS.

My strategy going into version 4 attempted to use modern EPC and cache the result, so that position changes due to CSS wouldn't interfere with the core algorithms depending on element position.

The problem is that managing a cache introduces a host of problems, and something as common as resizing your browser would completely invalidate the cache; furthermore, ScrollReveal has built up state by this point, and there's no easy way to tear everything down for another EPC snapshot to cache (without trampling the UX).

I'm sure there is a solution, but I haven't yet uncovered one of reasonable simplicity to implement.
